third_party/spdk_defs/nvme_defs.h

```diff
@@ -239,6 +239,295 @@ struct SglDescriptor {
 } ABSL_ATTRIBUTE_PACKED;
 static_assert(sizeof(SglDescriptor) == 16);
 
+// NVMe Base Specification Figure 70 to Figure 75
+// https://nvmexpress.org/wp-content/uploads/NVM-Express-1_4-2019.06.10-Ratified.pdf
+union VsRegister {
+  uint32_t raw;
+  struct {
+    uint32_t ter : 8;   // indicates the tertiary version
+    uint32_t mnr : 8;   // indicates the minor version
+    uint32_t mjr : 16;  // indicates the major version
+                        // reserved for 1.1, 1.2, 1.2.1 Compliant Controllers
+  } bits;
+} ABSL_ATTRIBUTE_PACKED;
+static_assert(sizeof(union VsRegister) == 4);
+
+// NVMe Base Specification Figure 248
+// https://nvmexpress.org/wp-content/uploads/NVM-Express-1_4-2019.06.10-Ratified.pdf
+struct PowerState {
+  uint16_t mp;  // bits 15:00: maximum power
+  uint8_t reserved1;
+  uint8_t mps : 1;   // bit 24: max power scale
+  uint8_t nops : 1;  // bit 25: non-operational state
+  uint8_t reserved2 : 6;
+  uint32_t enlat;   // bits 63:32: entry latency in microseconds
+  uint32_t exlat;   // bits 95:64: exit latency in microseconds
+  uint8_t rrt : 5;  // bits 100:96: relative read throughput
+  uint8_t reserved3 : 3;
+  uint8_t rrl : 5;  // bits 108:104: relative read latency
+  uint8_t reserved4 : 3;
+  uint8_t rwt : 5;  // bits 116:112: relative write throughput
+  uint8_t reserved5 : 3;
+  uint8_t rwl : 5;  // bits 124:120: relative write latency
+  uint8_t reserved6 : 3;  // includes fields added in NVMe Revision 1.4
+  uint8_t reserved7[16];  // includes fields added in NVMe Revision 1.4
+} ABSL_ATTRIBUTE_PACKED;
+static_assert(sizeof(struct PowerState) == 32);
+
+// identify controller data nvme over fabrics-specific fields
+// NVMe over Fabrics Figure 28
+// https://nvmexpress.org/wp-content/uploads/NVMe_over_Fabrics_1_0_Gold_20160605-1.pdf
+struct IdentifyControllerNvmfSpecific {
+  uint32_t ioccsz;  // i/o queue command capsule supported size (16-byte uni
+  uint32_t iorcsz;  // i/o queue response capsule supported size (16-byte uni
+  uint16_t icdoff;  // In-capsule data offset (16-byte uni
+
+  struct {
+    // Controller model: \ref spdk_nvmf_ctrlr_model
+    uint8_t ctrlr_model : 1;
+    uint8_t reserved : 7;
+  } ctrattr;  // Controller attributes
+
+  uint8_t msdbd;  // Maximum SGL block descriptors (0 = no limit)
+  uint8_t reserved[244];
+} ABSL_ATTRIBUTE_PACKED;
+static_assert(sizeof(IdentifyControllerNvmfSpecific) == 256);
+
+constexpr uint8_t kCtrlrSnLen = 20;  // controller serial number length
```

Comment:
I'd prefer defining constants at the start of the namespace for organizational purposes but I'm interested in what others think is the best approach.

Reply:
that's a good point. My reasoning was that I wanted the constant to be close to the struct where it is used. However, if the constant will be used in multiple places then putting the constant near one usage of it no longer makes sense, so it'd makes sense to put it at the start of the namespace. And if the constant is only used in one place, then perhaps it doesn't make sense to have a constant at all.